sale_coupon_validity_notification/models/coupon.py

```diff
@@ -0,0 +1,36 @@
+# Copyright 2023 Tecnativa - Pilar Vargas
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import _, models
+
+
+class CouponCoupon(models.Model):
+    _inherit = "coupon.coupon"
+
+    def write(self, vals):
+        res = super().write(vals)
+        if vals.get("state", False) == "expired":
+            for coupon in self.filtered(
+                lambda r: r.state == "expired" and r.sales_order_id
```

Comment:
This should be done **before** the super call so we can check the previous state of the coupons and notify only on those that were valid and now are getting expired. You aren't filtering only generated coupons either.